src/bio/bio_dev_monitor.c

```diff
@@ -0,0 +1,442 @@
+	D_PRINT("Data Units Written:          ");
+	dprint_uint128_dec(health_page->data_units_written);
+	D_PRINT("\n");
+	D_PRINT("Host Read Commands:          ");
+	dprint_uint128_dec(health_page->host_read_commands);
+	D_PRINT("\n");
+	D_PRINT("Host Write Commands:         ");
+	dprint_uint128_dec(health_page->host_write_commands);
+	D_PRINT("\n");
+	D_PRINT("Controller Busy Time:        ");
+	dprint_uint128_dec(health_page->controller_busy_time);
+	D_PRINT(" minutes\n");
+	D_PRINT("Power Cycles:                ");
+	dprint_uint128_dec(health_page->power_cycles);
+	D_PRINT("\n");
+	D_PRINT("Power On Hours:              ");
+	dprint_uint128_dec(health_page->power_on_hours);
+	D_PRINT(" hours\n");
+	D_PRINT("Unsafe Shutdowns:	     ");
+	dprint_uint128_dec(health_page->unsafe_shutdowns);
+	D_PRINT("\n");
+	D_PRINT("Unrecoverable Media Errors:  ");
+	dprint_uint128_dec(health_page->media_errors);
+	D_PRINT("\n");
+	D_PRINT("Lifetime Error Log Entries:  ");
+	dprint_uint128_dec(health_page->num_error_info_log_entries);
+	D_PRINT("\n");
+	D_PRINT("Warning Temperature Time:    %u minutes\n",
+		health_page->warning_temp_time);
+	D_PRINT("Critical Temperature Time:   %u minutes\n",
+		health_page->critical_temp_time);
+
+	/* Prep NVMe command to get controller data */
+	ctrlr_page_sz = sizeof(struct spdk_nvme_ctrlr_data);
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_IDENTIFY;
+	cmd.cdw10 = SPDK_NVME_IDENTIFY_CTRLR;
+
+	/*
+	 * Submit an NVMe Admin command to get controller data
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(health_desc, channel, &cmd,
+					   dev_health->bhm_ctrlr_buf,
+					   ctrlr_page_sz,
+					   get_spdk_identify_ctrlr_completion,
+					   dev_health);
+	if (rc)
+		D_ERROR("NVMe admin passthru (identify ctrlr), rc:%d\n", rc);
+}
+
+/* Get the SPDK device health state log and print all useful stats */
+void
+query_spdk_health_stats(struct bio_xs_context *ctxt, uint64_t now)
+{
+	struct bio_health_monitoring	*dev_health;
+	struct spdk_bdev		*bdev;
+	struct spdk_io_channel		*channel;
+	struct spdk_bdev_desc		*health_desc;
+	struct spdk_nvme_cmd		 cmd;
+	uint64_t			 health_stat_age;
+	int				 rc;
+	uint32_t			 numd, numdl, numdu;
+	uint32_t			 health_page_sz;
+
+	D_ASSERT(ctxt != NULL);
+	D_ASSERT(ctxt->bxc_blobstore != NULL);
+
+	dev_health = ctxt->bxc_blobstore->bb_dev_health;
+	health_stat_age = dev_health->bhm_stat_age;
+	/*
+	 * TODO Decide on an appropriate period to query device health
+	 * stats. Currently set at 1 min.
+	 */
+	if (health_stat_age + DAOS_SPDK_STATS_PERIOD >= now)
+		return;
+	dev_health->bhm_stat_age = now;
+
+
+	if (!dev_health->bhm_desc)
+		return;
+	health_desc = dev_health->bhm_desc;
+
+	bdev = spdk_bdev_desc_get_bdev(health_desc);
+	if (bdev == NULL) {
+		D_ERROR("Failed to get bdev from health desc:%p\n",
+			health_desc);
+		return;
+	}
+
+	if (!spdk_bdev_io_type_supported(bdev, SPDK_BDEV_IO_TYPE_NVME_ADMIN)) {
+		D_ERROR("Bdev NVMe admin passthru not supported!\n");
```

Comment:
It's not an error, otherwise, surplus error log will be generated by AIO or Malloc device.

Reply:
I think debug log not being printed by default will be better, otherwise, this message will flood the debug log.